Mathlib/RingTheory/Valuation/ValExtension.lean

```diff
@@ -0,0 +1,181 @@
+    (h : vA.integer.comap (algebraMap K A) = vK.integer) : IsValExtension vK vA where
+  val_isEquiv_comap := by
+    rw [isEquiv_iff_val_le_one]
+    intro x
+    rw [← Valuation.mem_integer_iff, ← Valuation.mem_integer_iff]
+    rw [← h]
+    rfl
+
+end mk'
+
+section integer
+
+variable {R A ΓR ΓA : Type*} [CommRing R] [Ring A]
+    [LinearOrderedCommGroupWithZero ΓR] [LinearOrderedCommGroupWithZero ΓA]
+    [Algebra R A] {vR : Valuation R ΓR} {vA : Valuation A ΓA} [IsValExtension vR vA]
+
+instance integerAlgebra : Algebra vR.integer vA.integer where
+    smul r a := {
+      val := r • a,
+      property := by
+        rw [Valuation.mem_integer_iff,
+          show r • ↑a = algebraMap R A r * a by exact (Algebra.smul_def r (a : A))]
+        norm_num
+        apply mul_le_one'
+        · rw [val_map_le_one_iff vR]
+          exact r.2
+        · exact a.2
+    }
+    toFun r := {
+      val := algebraMap R A r,
+      property := by
+        simp only [Valuation.mem_integer_iff]
+        rw [val_map_le_one_iff vR]
+        exact r.2
+    }
+    map_one' := by
+      ext
+      simp
+    map_mul' _ _ := by
+      ext
+      simp
+    map_zero' := by
+      ext
+      simp
+    map_add' _ _ := by
+      ext
+      simp
+    commutes' _ _ := by
+      ext
+      exact Algebra.commutes _ _
+    smul_def' _ _ := by
+      ext
+      exact Algebra.smul_def _ _
+
+@[simp, norm_cast]
+theorem coe_algebraMap_integer (r : vR.integer) :
+    ((algebraMap vR.integer vA.integer) r : A) = (algebraMap R A) (r : R) := by
+  rfl
+
+instance instIsScalarTowerInteger : IsScalarTower vR.integer vA.integer A where
+  smul_assoc x y z := by
+    simp only [Algebra.smul_def]
+    exact mul_assoc _ _ _
+
+theorem integerAlgebra_injective {K A ΓK ΓA : Type*} [Field K] [Ring A] [Nontrivial A]
+    [LinearOrderedCommGroupWithZero ΓK] [LinearOrderedCommGroupWithZero ΓA]
+    [Algebra K A] (vK : Valuation K ΓK) (vA : Valuation A ΓA) [IsValExtension vK vA] :
+    Function.Injective (algebraMap vK.integer vA.integer) := by
+  intro x y h
+  simp only [Subtype.ext_iff, coe_algebraMap_integer] at h
+  ext
+  apply RingHom.injective (algebraMap K A) h
+
+instance instIsLocalRingHomValuationInteger {K L ΓK ΓL : Type*} [Field K] [Field L]
+    [LinearOrderedCommGroupWithZero ΓK] [LinearOrderedCommGroupWithZero ΓL]
+    [Algebra K L] {vK : Valuation K ΓK} {vL : Valuation L ΓL} [IsValExtension vK vL] :
+    IsLocalRingHom (algebraMap vK.integer vL.integer) where
+  map_nonunit r hr := by
+    by_cases h : r = 0
+    · simp [h] at hr
+    · apply Valuation.Integers.isUnit_of_one (v := vK)
+      · exact Valuation.integer.integers (v := vK)
+      · simp only [isUnit_iff_ne_zero, ne_eq]
+        exact Subtype.ext_iff.not.mp h
+      · apply Valuation.Integers.one_of_isUnit (Valuation.integer.integers (v := vL)) at hr
+        change vL (((algebraMap vK.integer vL.integer) r) : L) = 1 at hr
+        norm_cast at hr
+        rw [val_map_eq_one_iff vK] at hr
+        exact hr
```

Comment:
This works for arbitrary rings as long as the map `K -> L` is also a local ring hom.
```lean
instance instIsLocalRingHomValuationInteger {K L ΓK ΓL : Type*} [CommRing K] [CommRing L]
    [LinearOrderedCommGroupWithZero ΓK] [LinearOrderedCommGroupWithZero ΓL]
    [Algebra K L] [IsLocalRingHom (algebraMap K L)] {vK : Valuation K ΓK} {vL : Valuation L ΓL} [IsValExtension vK vL] :
    IsLocalRingHom (algebraMap vK.integer vL.integer) where
  map_nonunit r hr := by
    apply (Valuation.integer.integers (v := vK)).isUnit_of_one
    · exact (isUnit_map_iff (algebraMap K L) _).mp (hr.map (algebraMap _ L))
    · apply (Valuation.integer.integers (v := vL)).one_of_isUnit at hr
      exact (val_map_eq_one_iff vK vL _).mp hr
```
Note that we might be missing this instance somewhere
```lean
instance (priority := 100) {K L} [Field K] [CommRing L] [Nontrivial L]
    (f : K →+* L) : IsLocalRingHom f where
  map_nonunit r hr := by simpa using hr.ne_zero
```

Reply:
Thank you. I will put this instance inside the file `Mathlib.RingTheory.Ideal.LocalRing`.